specs/design-black-box.md

```diff
@@ -72,10 +72,121 @@ specific ConfigMaps/Secrets.
+[argo-cd#fargo/engine](https://github.com/alexmt/argo-cd/tree/fargo/engine) and [flux#fargo](https://github.com/alexmt/flux/tree/fargo) 
+
+The GitOps engine API consists of three main packages:
+* The `utils` package. It consist of loosely coupled packages that implements K8S resource diffing, health assessment, etc
+* The `app` package that Application data structure definition and CRUD client interface.
+* The `engine` package that leverages `utils` and uses provided set of Applications as a configuration source.
+
+```
+gitops-engine
+|-- pkg
+|   |-- utils
+|   |   |-- diff   # provides Kubernetes resource diffing functionality
+|   |   |-- kube   # provides utility methods to interact with Kubernetes
+|   |   |-- lue    # provides utility methods to run Lua scripts
+|   |   |-- apps   # provides utility methods to manipulate Applications (e.g. start sync operation, wait for sync operation, force reconciliation)
+|   |   `-- health # provides Kubernetes resources health assessment
+|   |-- app
+|   |   |-- apis       # contains data structures that describe Application
+|   |   `-- client     # contains client that provides Application CRUD operations
+|   `-- engine     # the engine implementation
+```
+
+The engine is responsible for reconciliation only which includes:
+- Interacting with the Kubernetes API: load and maintain the cluster state; pushing required changes to the Kubernetes API.
+- Reconciliation logic: based on provides target resource definition find which resources should be updated/deleted/created.
+- Syncing logic: determine the order in which resources should be modified; features like sync hooks, waves, etc.
+
+The manifests generation is out of scope and should be implemented by the Engine consumer.
+
+### Engine API
+
+The engine API includes Application data structure/client and `Engine` golang interface that allows configuring reconciliation process:
+
+**Engine interface** - provides set of that allows updating reconciliation settings and subscribe to engine events.
+```golang
+type Engine interface {
+	// Run starts reconciliation loop using specified number of processors for reconciliation and operation execution.
+	Run(ctx context.Context, statusProcessors int, operationProcessors int)
+
+	// SetReconciliationSettings updates reconciliation settings
+	SetReconciliationSettings(settings ReconciliationSettings)
+
+	// OnBeforeSync registers callback that is executed before each sync operation.
+	OnBeforeSync(callback func(appName string, tasks []SyncTaskInfo) ([]SyncTaskInfo, error)) Unsubscribe
+
+	// OnSyncCompleted registers callback that is executed after each sync operation.
+	OnSyncCompleted(callback func(appName string, state appv1.OperationState) error) Unsubscribe
+
+	// OnClusterCacheInitialized registers that that is executed when cluster cache initialization is completed.
+	OnClusterCacheInitialized(callback func(server string)) Unsubscribe
```

Comment:
Why is this needed?

Reply:
Judging by the changes in the `fargo` branch of Flux, it seems to be used to kickstart the engine. https://github.com/alexmt/flux/commit/b4cc9b6a99966f75a03687a1b0599b1e8b5ef047#diff-6b48cc88be8e3f8560bbe695aa8b202fR90

Reply:
@hiddeco  is right. The callback is required to block `RunEngine` until cache is initialized and to make sure that Flux don't use the incomplete list of workloads.

Reply:
done. Added comment.